datafusion/substrait/tests/roundtrip_logical_plan.rs

```diff
@@ -278,6 +278,18 @@ mod tests {
         roundtrip("SELECT a,b,c,d,e FROM datafusion.public.data;").await
     }
 
+    #[tokio::test]
+    async fn roundtrip_inner_join_table_reuse() -> Result<()> {
```

Comment:
Given there are offsets it might help to also test columns other than `a` (which is at offset 0 in each schema)

Perhaps separate tests for the following cases would add better coverage:

* `SELECT d1.b, d2.c FROM data d1 JOIN data d2 ON d1.b = d2.b`
* `SELECT d1.b, d2.c FROM data d1 JOIN data d2 ON d1.b = d2.c`

Reply:
Also as I recall `USING` is also special so maybe a test for:

* `SELECT d1.b, d2.c FROM data d1 JOIN data d2 using (b)`

Reply:
Thanks, @alamb!

I added the first test case.

For the second, since `c` is type `DATE`, the join condition is invalid. So I tried:
```sql
SELECT d1.b, d2.c FROM data d1 JOIN data d2 ON CAST(d1.b AS int) = d2.e
```
and got an error due to unsupported expression in join condition. I filed an issue to track [here](https://github.com/apache/arrow-datafusion/issues/6544).


For the `using` case, this requires producer support for the join constraint so I got the error `Error: NotImplemented("join constraint")`. Let me know if you want this issue filed as well.